modepy/nodes.py

```diff
@@ -473,46 +483,75 @@ def _random_nodes_for_simplex(shape: Simplex, nnodes: int, rng=None):
 # }}}
 
 
-# {{{ QN
+# {{{ generic tensor product space
 
-@node_tuples_for_space.register(QN)
-def _node_tuples_for_qn(space: QN):
-    from pytools import \
-            generate_nonnegative_integer_tuples_below as gnitb
-    return tuple([
-        tp[::-1] for tp in gnitb(space.order + 1, space.spatial_dim)
+@node_tuples_for_space.register(TensorProductSpace)
+def _node_tuples_for_tp(space: TensorProductSpace):
+    from pytools import generate_nonnegative_integer_tuples_below as gnitb
+    tuples_for_space = [node_tuples_for_space(s) for s in space.bases]
+
+    n = len(tuples_for_space)
+    r = tuple([
+        sum((tuples_for_space[n - i - 1][j] for i, j in enumerate(tp[::-1])), ())
+        for tp in gnitb([len(tp) for tp in tuples_for_space])
         ])
 
+    return r
+
 
-@equispaced_nodes_for_space.register(QN)
-def _equispaced_nodes_for_qn(space: QN, shape: Hypercube):
-    if not isinstance(shape, Hypercube):
+@equispaced_nodes_for_space.register(TensorProductSpace)
+def _equispaced_nodes_for_tp(
+        space: TensorProductSpace,
+        shape: TensorProductShape):
+    if not isinstance(shape, TensorProductShape):
         raise NotImplementedError((type(space).__name__, type(shape).__name__))
+
     if space.spatial_dim != shape.dim:
         raise ValueError("spatial dimensions of shape and space must match")
 
-    if space.order == 0:
-        return np.array(node_tuples_for_space(space), dtype=np.float64).T
-    else:
-        return (np.array(node_tuples_for_space(space), dtype=np.float64)
-                / space.order*2 - 1).T
+    return tensor_product_nodes([
+        equispaced_nodes_for_space(b, s)
+        for b, s in zip(space.bases, shape.bases)
+        ])
 
 
-@edge_clustered_nodes_for_space.register(QN)
-def _edge_clustered_nodes_for_qn(space: QN, shape: Hypercube):
-    if not isinstance(shape, Hypercube):
+@edge_clustered_nodes_for_space.register(TensorProductSpace)
+def _edge_clustered_nodes_for_tp(
+        space: TensorProductSpace,
+        shape: TensorProductShape):
+    if not isinstance(shape, TensorProductShape):
         raise NotImplementedError((type(space).__name__, type(shape).__name__))
+
     if space.spatial_dim != shape.dim:
         raise ValueError("spatial dimensions of shape and space must match")
 
-    return legendre_gauss_lobatto_tensor_product_nodes(
-            space.spatial_dim, space.order)
+    return tensor_product_nodes([
+        edge_clustered_nodes_for_space(b, s)
+        for b, s in zip(space.bases, shape.bases)
+        ])
+
+
+@random_nodes_for_shape.register(TensorProductShape)
+def _random_nodes_for_tp(shape: TensorProductShape, nnodes: int, rng=None):
+    if rng is None:
+        rng = np.random.default_rng()
+
+    # FIXME: this may not be exactly uniform per dimension; any better way?
```

Comment:
I'm not sure why you're tensor-producting randomly drawn nodes on each sub-shape. That'll come out in a grid, i.e. not very random. It'll also generate a different number of nodes than the user asked for.

Why not have each dimension generate the full `nnodes` and then "zip" them together?

Reply:
Hm, yeah, not quite sure why I thought that wouldn't work.

Cleaned up in 0911d70.